qualang_tools/addons/calibration/calibrations_API.py

```diff
@@ -0,0 +1,1114 @@
+"""
+CALIBRATIONS
+"""
+from qm.qua import *
+from qualang_tools.plot import interrupt_on_close
+from qualang_tools.results import progress_counter, fetching_tool
+from qualang_tools.units import unit
+from qualang_tools.loops import from_array
+from qm import SimulationConfig
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy import signal
+
+
+available_variables = ["frequency", "amplitude", "duration"]
+u = unit()
+
+
+class QUA_calibrations:
+    def _check_element(self, element):
+        if element in self.config["elements"]:
+            return True
+        else:
+            raise Exception(f"The element '{element}' is not in the current config.")
+
+    def _check_operation(self, operation, element):
+        if operation in self.config["elements"][element]["operations"]:
+            return True
+        else:
+            raise Exception(
+                f"The operation '{operation}' is not in the current config for element '{element}'."
+            )
+
+    def _check_iw(self, int_weights):
+        for iw in int_weights:
+            if (
+                iw
+                not in self.config["pulses"][
+                    self.config["elements"][self.readout_elmt]["operations"][
+                        self.readout_op
+                    ]
+                ]["integration_weights"]
+            ):
+                raise Exception(
+                    f"The integration weight '{iw}' is not in the current config for operation '{self.readout_op}'."
+                )
+        return True
+
+    def _check_outputs(self, outputs):
+        for out in outputs:
+            if out not in self.config["elements"][self.readout_elmt]["outputs"]:
+                raise Exception(
+                    f"The output '{out}' is not in the current config for element '{self.readout_elmt}'."
+                )
+        return True
+
+    def _check_plot_options(self, plot_options):
+        if type(plot_options) is not dict:
+            raise Exception("plot_options must be a python dictionary.")
+        for key in plot_options:
+            if key in self.plot_options.keys():
+                self.plot_options[key] = plot_options[key]
+            else:
+                raise Exception(
+                    f"Plot option {key} is not implemented, available options are {self.plot_options.keys()}"
+                )
+
+    def _user_checks(self, calibration, scan_variables, iterations, cooldown_time):
+        self.calibration_list.append(calibration)
```

Comment:
What happens if a user adds the same calibration twice? I think it would break things... It would be added to the list twice, but the variables would be overridden and only used once.

Reply:
Correct, I didn't think of this case...